syntax = "proto3";

package sancus.institution;

import "protobufs/webauthn.proto";
import "protobufs/account.proto";

service Exchange {
    rpc InitiateExchange(InitiateExchangeRequest) returns (InitiateExchangeResponse);
    rpc ProcessExchange(ProcessExchangeRequest) returns (ProcessExchangeResponse);
}

message InitiateExchangeRequest {
    bytes fromAccountId = 1;
    bytes toAccountId = 2;
    string amount = 3;
    string exchangeRateJWT = 4;
}

message InitiateExchangeResponse {
    bytes id = 1;
    ChallengeRequest challengeRequest = 2;  // challenge is of type AccountDeltaGroupChallengeRequest
    repeated RevealedAccountDeltaCommitment revealedCommitments = 3;
    PublicKeyCredentialRequestOptions credentialRequest = 4;
}

message ProcessExchangeRequest {
    bytes id = 1;
    AuthenticatorAssertionResponse assertion = 2;
}

message ProcessExchangeResponse {
}
